File: sender/java/src/main/java/org/file/qrcode/protocol/RobustSoliton.java
package org.file.qrcode.protocol;

/**
 * Robust Soliton 度分布（CONTRACT.md 4.1 / 4.2）。
 *
 * 关键点：ln 结果量化到 2^-16 网格（lnq），把跨平台 libm 的 ulp 级差异挡在网格之外；
 * 其余全部是 IEEE 双精度加减乘除，逐位确定。CDF 量化成 uint32，用 long 承载
 * （末档是 2^32，放不进 uint32），抽样全程整数比较。
 */
public final class RobustSoliton {

    public static final double C = 0.03;
    public static final double DELTA = 0.05;
    private static final double LN_GRID = 65536.0;
    /** 2^32，末档兜底值 */
    public static final long TWO_POW_32 = 4294967296L;

    private final int K;
    /** 长度 K+1，下标 0 不用 */
    private final long[] cdf;

    public RobustSoliton(int K) {
        if (K < 1) {
            throw new IllegalArgumentException("K 必须 >= 1");
        }
        this.K = K;
        this.cdf = buildCdf(K, C, DELTA);
    }

    /** 量化对数：floor(ln(x) * 65536 + 0.5) / 65536 */
    static double lnq(double x) {
        return Math.floor(Math.log(x) * LN_GRID + 0.5) / LN_GRID;
    }

    /** 按契约 4.2 计算量化 CDF，返回长度 K+1 的数组。 */
    static long[] buildCdf(int K, double c, double delta) {
        double[] rho = new double[K + 2];
        rho[1] = 1.0 / K;
        for (int d = 2; d <= K; d++) {
            rho[d] = 1.0 / ((double) d * (d - 1));
        }

        double R = c * lnq(K / delta) * Math.sqrt(K);
        double[] tau = new double[K + 2];
        int kr = (int) (K / R);
        if (kr < 1) {
            kr = 1;
        }
        int lim = Math.min(kr, K + 1);
        for (int d = 1; d < lim; d++) {
            tau[d] = R / ((double) d * K);
        }
        if (kr <= K) {
            double v = R * lnq(R / delta) / K;
            tau[kr] = v > 0.0 ? v : 0.0;
        }

        // 升序累加，顺序不可改
        double Z = 0.0;
        for (int d = 1; d <= K; d++) {
            Z += rho[d] + tau[d];
        }

        long[] cdf = new long[K + 1];
        double acc = 0.0;
        for (int d = 1; d <= K; d++) {
            acc += (rho[d] + tau[d]) / Z;
            long q = (long) (acc * 4294967296.0);
            if (q > 0xFFFFFFFFL) {
                q = TWO_POW_32;
            }
            cdf[d] = q;
        }
        cdf[K] = TWO_POW_32;
        return cdf;
    }

    public int getK() {
        return K;
    }

    /** 取度数 d 对应的量化 CDF 值，d 从 1 起。 */
    public long cdfAt(int d) {
        return cdf[d];
    }

    /**
     * 抽度数。先推进 state，再在 CDF 上二分。
     *
     * Java 无多返回值，把结果打包进一个 long：高 32 位是度数，低 32 位是推进后的 state。
     */
    public long sampleDegree(int state) {
        int next = Prng.xorshift32(state);
        long r = next & 0xFFFFFFFFL;
        int lo = 1;
        int hi = K;
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (cdf[mid] <= r) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return ((long) lo << 32) | (next & 0xFFFFFFFFL);
    }
}

File: receiver/QrDrop/Protocol/RobustSoliton.swift
//
//  RobustSoliton.swift
//  QrDrop
//
//  Robust Soliton 度分布与量化 CDF。严格对应 CONTRACT.md 第 4.1 / 4.2 节。
//  ln 结果量化到 2^-16 网格，避免跨平台 libm 的 ulp 差异传导到度数抽样。
//

import Foundation

struct RobustSoliton {

    static let defaultC = 0.03
    static let defaultDelta = 0.05
    private static let lnGrid = 65536.0
    /// 2^32，超出 UInt32 范围，故 CDF 用 UInt64 存
    static let cdfTop: UInt64 = 4_294_967_296

    let K: Int
    /// 长度 K+1，下标 0 不用；cdf[K] 恒为 2^32
    let cdf: [UInt64]

    /// 量化对数：floor(ln(x) * 65536 + 0.5) / 65536
    @inline(__always)
    static func lnq(_ x: Double) -> Double {
        (Foundation.log(x) * lnGrid + 0.5).rounded(.down) / lnGrid
    }

    init(K: Int, c: Double = RobustSoliton.defaultC, delta: Double = RobustSoliton.defaultDelta) {
        self.K = K
        let kd = Double(K)

        var rho = [Double](repeating: 0, count: K + 2)
        rho[1] = 1.0 / kd
        if K >= 2 {
            for d in 2...K {
                rho[d] = 1.0 / (Double(d) * Double(d - 1))
            }
        }

        let R = c * RobustSoliton.lnq(kd / delta) * kd.squareRoot()
        var tau = [Double](repeating: 0, count: K + 2)
        let kr = max(1, Int(kd / R))
        // 注意上界：与参考实现的 range(1, min(kr, K+1)) 一致，取不到上界本身
        let upper = min(kr, K + 1)
        if upper > 1 {
            for d in 1..<upper { tau[d] = R / (Double(d) * kd) }
        }
        if kr <= K {
            tau[kr] = max(0.0, R * RobustSoliton.lnq(R / delta) / kd)
        }

        // 归一化因子必须按 d 升序累加，顺序影响浮点结果
        var Z = 0.0
        for d in 1...K { Z += rho[d] + tau[d] }

        var table = [UInt64](repeating: 0, count: K + 1)
        var acc = 0.0
        for d in 1...K {
            acc += (rho[d] + tau[d]) / Z
            let v = acc * 4_294_967_296.0
            table[d] = v >= 4_294_967_296.0 ? RobustSoliton.cdfTop : UInt64(v)
        }
        table[K] = RobustSoliton.cdfTop      // 末档兜底，保证一定命中
        self.cdf = table
    }

    /// 抽度数：推进状态后在 CDF 上二分，返回 (度数, 新状态)
    @inline(__always)
    func sampleDegree(state: UInt32) -> (degree: Int, state: UInt32) {
        let s = Prng.xorshift32(state)
        let r = UInt64(s)
        var lo = 1
        var hi = K
        while lo < hi {
            let mid = (lo + hi) / 2
            if cdf[mid] <= r { lo = mid + 1 } else { hi = mid }
        }
        return (lo, s)
    }
}
